algorithms/CPlusPlus/Trees/binary-search-tree.cpp

```diff
@@ -47,6 +48,118 @@ void Insert(TreeNode*& root, const int& data){
 	else 			    { Insert(root->left, data); }
 }
 
+
+bool isfull(TreeNode* root){
+        /**
+         *
+         * Check if a binary tree is full or not
+         * A binary tree is full when every node in the 
+         * tree has either two or zero child nodes.
+         *
+         * @params: `root` root/parent node of the tree
+         *
+         * @return: true if it the binary tree is full else false
+         */
+
+        if(root == nullptr)                                     { return true; }
+        if(root->left == nullptr && root->right == nullptr)     { return true; }
+        if((root->left != nullptr && root->right != nullptr) )  { return isfull(root->left) && isfull(root->right); }
+        return false;
+}
+
+int depth(TreeNode* root){
+        /**
+         * Find the depth of the left most tree.
+         * Here the depth of the left most tree is found but 
+         * it is only a matter of preference.
+         *
+         * @params: `root` root/parent node of the tree
+         *
+         * @return: `d` returns the depth of the left most tree
+         */
+
+        int d = 0;
+        while(root != nullptr){
+                root = root->left;
+                d++;
+        }
+        return d;
+}
+
+
+bool perfect_recursive(TreeNode* cur, int depth, int level = 0){
+        /**
+         * A Recursive strategy to check if a tree is perfect or not
+         *
+         * A binary tree is perfect if when all the inner node's
+         * has two children and the all the leaf node's are at the
+         * same level.
+         *
+         * @params: `cur`  node of the tree
+         * @params: `depth` depth of the left most tree
+         * @params: `level` level of the cur node
+         *
+         * @return: true if the binary tree is perfect else false
+         */
+
+        if(cur == nullptr)                                { return true; }
+        if(cur->left == nullptr && cur->right == nullptr) { return depth == level; }
+        if(cur->left == nullptr || cur->right == nullptr) { return false; }
+        if(cur->left != nullptr && cur->right != nullptr) {
+                return perfect_recursive(cur->left, depth, level+1) && perfect_recursive(cur->right, depth, level+1);
+        }
+}
+
+int count_nodes(TreeNode* cur){
+        /**
+         * Count the number of node's in the tree
+         * @params: `cur`  node of the tree
+         */
+
+        if(cur != nullptr){ return 1 + count_nodes(cur->left) + count_nodes(cur->right); }
```

Comment:
What if `cur` is nullptr ? 

you are lucky that gcc handles it for you by sending a zero in this case but every compiler is not gcc ( or similar ).

it's recommended to be prepare for worse and avoid letting your compiler take decision for you, and put a `return 0` here in any other case. 
```suggestion
        if(cur != nullptr){ return 1 + count_nodes(cur->left) + count_nodes(cur->right); }
        return 0;
```